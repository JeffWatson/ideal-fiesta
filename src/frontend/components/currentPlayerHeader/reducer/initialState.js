// TODO remove from here. Shouldn't exist. Move to tests only location
import { fromJS } from 'immutable';

export default fromJS({
  enabled: false,
  currentPlayer: {
    username: 'bananaMan',
    avatarUrl: './favicon.ico',
    status: {
      money: 1000,
      specialPoints: 90,
    },
  },
});
