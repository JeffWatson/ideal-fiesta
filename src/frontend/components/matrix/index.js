import MatrixView from './matrix';

export { MatrixView as default };
