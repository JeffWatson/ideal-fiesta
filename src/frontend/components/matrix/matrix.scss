@import "../../app/style/variables";

.matrix {
  .matrix-table {
    .matrix-cell {
      height: $matrixCellHeight;
      width: $matrixCellWidth;
      border: 3px solid $matrix-unselected-color;

      &.movable {
        @include matrix-container($matrix-movable-color);
        border: 3px solid $matrix-movable-color;
      }

      &.attackable:not(.movable) {
        @include matrix-container($matrix-attackable-color);
        border: 3px solid $matrix-attackable-color;
      }

      &.selected {
        @include matrix-container($matrix-selected-color);
        border: 3px solid $matrix-selected-color;
      }

      &.actionable {
        @include matrix-container($matrix-actionable-color);
        border: 3px solid $matrix-actionable-color;
      }

      & > div {
        height: $matrixCellHeight;
        width: $matrixCellWidth;
        z-index: -1;
      }
    }
  }
}