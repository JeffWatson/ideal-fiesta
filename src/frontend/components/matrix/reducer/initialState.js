import { fromJS } from 'immutable';


// TODO need to refactor units... they should have health. Buildings too.
const initialState = {
  matrix: {
    columns: 5,
    rows: 5,
    grid: [
      [{ terrain: 'SAND', unit: 'SOLDIER', health: 20 },
        { terrain: 'SAND' },
        { terrain: 'SAND' },
        { terrain: 'SAND', unit: 'SMALL_TANK', health: 20 },
        { terrain: 'SAND' }],
      [{ terrain: 'OCEAN' },
        { terrain: 'OCEAN' },
        { terrain: 'OCEAN' },
        { terrain: 'OCEAN' },
        { terrain: 'LAND_FACTORY', health: 20 }],
      [{ terrain: 'GRASS', unit: 'SOLDIER', health: 20 },
        { terrain: 'SAND' },
        { terrain: 'MOUNTAIN' },
        { terrain: 'SAND' },
        { terrain: 'SAND' }],
      [{ terrain: 'SAND' },
        { terrain: 'SAND' },
        { terrain: 'SAND' },
        { terrain: 'GRASS', unit: 'SMALL_TANK', health: 20 },
        { terrain: 'SAND' }],
      [{ terrain: 'SAND' },
        { terrain: 'GRASS' },
        { terrain: 'SAND' },
        { terrain: 'SAND' },
        { terrain: 'SAND', unit: 'SOLDIER', health: 20 }],
    ],
  },
};

export default fromJS(initialState);
