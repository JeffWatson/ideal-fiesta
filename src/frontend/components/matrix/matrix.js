import React, { Component } from 'react';
import { func } from 'prop-types';
import { connect } from 'react-redux';
import { times } from 'lodash';
import classNames from 'classnames';
import UnitFactory from './../units/unitFactory';
import TerrainFactory from './../terrain/terrainFactory';
import mapStateToProps from './selector/mapStateToProps';
import mapDispatchToProps from './actions/mapDispatchToProps';

import './matrix.scss';

class MatrixView extends Component {
  generateMatrix({ matrix, currentPlayer }) {
    return (times(matrix.get('rows'), row => (<tr className="matrix-row" key={`matrix-row-${row}`}>
      {times(matrix.get('columns'), (column) => {
        const cell = matrix.getIn(['grid', row, column]);
        return this.renderCell({ column, row, cell, currentPlayer });
      })}
    </tr>)));
  }

  renderCell({ column, row, cell, currentPlayer }) {
    const { onCellClick } = this.props;
    const health = cell.get('health');
    const unit = cell.get('unit');
    const terrain = cell.get('terrain');
    const selected = cell.get('selected');
    const attackable = cell.get('attackable');
    const movable = cell.get('movable');
    const actionable = cell.get('actionable');
    const player = cell.get('player');
    const moveDirection = cell.get('moveDirection');
    const disabled = cell.get('disabled');
    const className = classNames('matrix-cell', { attackable, movable, actionable, selected });

    const key = `matrix-cell-${column}-${row}-${terrain}`;
    const children = UnitFactory.createUnit({ player, factoryType: 'LAND', unit, disabled });
    return (<td
      className={className}
      key={key}
      onClick={() => onCellClick({ terrain,
        column,
        row,
        unit,
        matrix: this.props.matrix,
        currentPlayer,
      })}
    >
      {TerrainFactory.createTerrain({ terrain, children, row, column, health, moveDirection })}
    </td>);
  }

  render() {
    const { matrix } = this.props;
    const currentPlayer = matrix.get('currentPlayer');

    return (<div className="matrix">
      <div className="current-player-info">The current Player is: { currentPlayer }</div>
      <table className="matrix-table">
        <tbody>
          {this.generateMatrix({ matrix, currentPlayer })}
        </tbody>
      </table>
    </div>);
  }
}

// TODO proper propTypes of matrix would be nice.
MatrixView.propTypes = {
  onCellClick: func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(MatrixView);
