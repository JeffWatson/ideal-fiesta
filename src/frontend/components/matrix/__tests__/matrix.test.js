import React from 'react';
import { shallow } from 'enzyme';
import { MatrixView } from './../matrix';
import initialState from './../reducer/initialState';

describe('frontend/components/matrix component', () => {
  it('renders correctly', () => {
    const component = shallow(<MatrixView matrix={initialState.get('matrix')} />);

    expect(component).toMatchSnapshot();
  });
});
