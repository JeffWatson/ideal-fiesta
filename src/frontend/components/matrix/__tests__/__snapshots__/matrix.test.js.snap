// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`frontend/components/matrix component CONNECTED selects initial state properly 1`] = `
<MatrixView
  matrix={
    Immutable.Map {
      currentPlayer: "BLUE",
      columns: 5,
      rows: 5,
      players: Immutable.List [
        "BLUE",
        "GREEN",
        "RED",
        "YELLOW",
      ],
      grid: Immutable.List [
        Immutable.List [
          Immutable.Map {
            terrain: "SAND",
            unit: Immutable.Map {
              actor: "SOLDIER",
              health: 20,
              player: "GREEN",
            },
          },
          Immutable.Map {
            terrain: "SAND",
          },
          Immutable.Map {
            terrain: "SAND",
          },
          Immutable.Map {
            terrain: "SAND",
            unit: Immutable.Map {
              actor: "SMALL_TANK",
              health: 20,
              player: "RED",
            },
          },
          Immutable.Map {
            terrain: "SAND",
          },
        ],
        Immutable.List [
          Immutable.Map {
            terrain: "OCEAN",
          },
          Immutable.Map {
            terrain: "OCEAN",
          },
          Immutable.Map {
            terrain: "OCEAN",
          },
          Immutable.Map {
            terrain: "OCEAN",
          },
          Immutable.Map {
            terrain: "LAND_FACTORY",
            building: Immutable.Map {
              health: 20,
              player: "RED",
            },
          },
        ],
        Immutable.List [
          Immutable.Map {
            terrain: "GRASS",
          },
          Immutable.Map {
            terrain: "SAND",
          },
          Immutable.Map {
            terrain: "MOUNTAIN",
          },
          Immutable.Map {
            terrain: "GRASS",
            unit: Immutable.Map {
              actor: "SOLDIER",
              health: 20,
              player: "BLUE",
            },
          },
          Immutable.Map {
            terrain: "SAND",
          },
        ],
        Immutable.List [
          Immutable.Map {
            terrain: "LAND_FACTORY",
            building: Immutable.Map {
              health: 20,
              player: "BLUE",
            },
          },
          Immutable.Map {
            terrain: "SAND",
          },
          Immutable.Map {
            terrain: "SAND",
          },
          Immutable.Map {
            terrain: "GRASS",
            unit: Immutable.Map {
              actor: "SOLDIER",
              health: 20,
              player: "BLUE",
            },
          },
          Immutable.Map {
            terrain: "SAND",
          },
        ],
        Immutable.List [
          Immutable.Map {
            terrain: "SAND",
          },
          Immutable.Map {
            terrain: "GRASS",
          },
          Immutable.Map {
            terrain: "SAND",
            unit: Immutable.Map {
              actor: "SMALL_TANK",
              health: 20,
              player: "BLUE",
            },
          },
          Immutable.Map {
            terrain: "SAND",
          },
          Immutable.Map {
            terrain: "SAND",
            unit: Immutable.Map {
              actor: "SOLDIER",
              health: 20,
              player: "YELLOW",
            },
          },
        ],
      ],
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;

exports[`frontend/components/matrix component renders correctly 1`] = `
<div
  className="matrix"
>
  <div
    className="current-player-info"
  >
    <Connect(CurrentPlayerHeader)
      currentPlayer="BLUE"
    />
  </div>
  <table
    className="matrix-table"
  >
    <tbody>
      <tr
        className="matrix-row"
        key="matrix-row-0"
      >
        <td
          className="matrix-cell"
          key="matrix-cell-0-0-SAND"
        >
          <Connect(Terrain)
            column={0}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={true}
            isOccupiedByPlayer={false}
            row={0}
            terrain={
              Object {
                "defenseBonus": 1,
              }
            }
          >
            <Sand
              column={0}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={true}
              isOccupiedByPlayer={false}
              row={0}
              terrain={
                Object {
                  "defenseBonus": 1,
                }
              }
            >
              <div>
                <div>
                  <Soldier
                    attacks={
                      Array [
                        "AIR",
                        "LAND",
                        "SEA",
                      ]
                    }
                    disabled={false}
                    inaccessibleTerrain={
                      Array [
                        "OCEAN",
                      ]
                    }
                    name="SOLDIER"
                    player="green"
                    stats={
                      Object {
                        "attack": 0.4,
                        "defense": 4,
                        "move": 3,
                        "range": 1,
                      }
                    }
                    type="LAND"
                  />
                  <Health
                    className="unit"
                    health={20}
                  />
                </div>
              </div>
            </Sand>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-1-0-SAND"
        >
          <Connect(Terrain)
            column={1}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={0}
            terrain={
              Object {
                "defenseBonus": 1,
              }
            }
          >
            <Sand
              column={1}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={0}
              terrain={
                Object {
                  "defenseBonus": 1,
                }
              }
            >
              <div />
            </Sand>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-2-0-SAND"
        >
          <Connect(Terrain)
            column={2}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={0}
            terrain={
              Object {
                "defenseBonus": 1,
              }
            }
          >
            <Sand
              column={2}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={0}
              terrain={
                Object {
                  "defenseBonus": 1,
                }
              }
            >
              <div />
            </Sand>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-3-0-SAND"
        >
          <Connect(Terrain)
            column={3}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={true}
            isOccupiedByPlayer={false}
            row={0}
            terrain={
              Object {
                "defenseBonus": 1,
              }
            }
          >
            <Sand
              column={3}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={true}
              isOccupiedByPlayer={false}
              row={0}
              terrain={
                Object {
                  "defenseBonus": 1,
                }
              }
            >
              <div>
                <div>
                  <SmallTank
                    attacks={
                      Array [
                        "AIR",
                        "LAND",
                        "SEA",
                      ]
                    }
                    disabled={false}
                    inaccessibleTerrain={
                      Array [
                        "OCEAN",
                        "MOUNTAIN",
                      ]
                    }
                    name="SMALL_TANK"
                    player="red"
                    stats={
                      Object {
                        "attack": 0.8,
                        "defense": 6,
                        "move": 4,
                        "range": 1,
                      }
                    }
                    type="LAND"
                  />
                  <Health
                    className="unit"
                    health={20}
                  />
                </div>
              </div>
            </Sand>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-4-0-SAND"
        >
          <Connect(Terrain)
            column={4}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={0}
            terrain={
              Object {
                "defenseBonus": 1,
              }
            }
          >
            <Sand
              column={4}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={0}
              terrain={
                Object {
                  "defenseBonus": 1,
                }
              }
            >
              <div />
            </Sand>
          </Connect(Terrain)>
        </td>
      </tr>
      <tr
        className="matrix-row"
        key="matrix-row-1"
      >
        <td
          className="matrix-cell"
          key="matrix-cell-0-1-OCEAN"
        >
          <Connect(Terrain)
            column={0}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={1}
            terrain={
              Object {
                "defenseBonus": 0,
              }
            }
          >
            <Ocean
              column={0}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={1}
              terrain={
                Object {
                  "defenseBonus": 0,
                }
              }
            >
              <div />
            </Ocean>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-1-1-OCEAN"
        >
          <Connect(Terrain)
            column={1}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={1}
            terrain={
              Object {
                "defenseBonus": 0,
              }
            }
          >
            <Ocean
              column={1}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={1}
              terrain={
                Object {
                  "defenseBonus": 0,
                }
              }
            >
              <div />
            </Ocean>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-2-1-OCEAN"
        >
          <Connect(Terrain)
            column={2}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={1}
            terrain={
              Object {
                "defenseBonus": 0,
              }
            }
          >
            <Ocean
              column={2}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={1}
              terrain={
                Object {
                  "defenseBonus": 0,
                }
              }
            >
              <div />
            </Ocean>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-3-1-OCEAN"
        >
          <Connect(Terrain)
            column={3}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={1}
            terrain={
              Object {
                "defenseBonus": 0,
              }
            }
          >
            <Ocean
              column={3}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={1}
              terrain={
                Object {
                  "defenseBonus": 0,
                }
              }
            >
              <div />
            </Ocean>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-4-1-LAND_FACTORY"
        >
          <Connect(Terrain)
            column={4}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            player="RED"
            row={1}
            terrain={
              Object {
                "actionable": true,
                "defenseBonus": 1,
                "selectable": true,
              }
            }
          >
            <LandFactory
              column={4}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              player="RED"
              row={1}
              terrain={
                Object {
                  "actionable": true,
                  "defenseBonus": 1,
                  "selectable": true,
                }
              }
            >
              <div>
                <Health
                  className="building"
                  health={20}
                />
              </div>
            </LandFactory>
          </Connect(Terrain)>
        </td>
      </tr>
      <tr
        className="matrix-row"
        key="matrix-row-2"
      >
        <td
          className="matrix-cell"
          key="matrix-cell-0-2-GRASS"
        >
          <Connect(Terrain)
            column={0}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={2}
            terrain={
              Object {
                "defenseBonus": 0,
              }
            }
          >
            <Grass
              column={0}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={2}
              terrain={
                Object {
                  "defenseBonus": 0,
                }
              }
            >
              <div />
            </Grass>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-1-2-SAND"
        >
          <Connect(Terrain)
            column={1}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={2}
            terrain={
              Object {
                "defenseBonus": 1,
              }
            }
          >
            <Sand
              column={1}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={2}
              terrain={
                Object {
                  "defenseBonus": 1,
                }
              }
            >
              <div />
            </Sand>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-2-2-MOUNTAIN"
        >
          <Connect(Terrain)
            column={2}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={2}
            terrain={
              Object {
                "defenseBonus": 2,
              }
            }
          >
            <Mountain
              column={2}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={2}
              terrain={
                Object {
                  "defenseBonus": 2,
                }
              }
            >
              <div />
            </Mountain>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-3-2-GRASS"
        >
          <Connect(Terrain)
            column={3}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={true}
            isOccupiedByPlayer={true}
            row={2}
            terrain={
              Object {
                "defenseBonus": 0,
              }
            }
          >
            <Grass
              column={3}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={true}
              isOccupiedByPlayer={true}
              row={2}
              terrain={
                Object {
                  "defenseBonus": 0,
                }
              }
            >
              <div>
                <div>
                  <Soldier
                    attacks={
                      Array [
                        "AIR",
                        "LAND",
                        "SEA",
                      ]
                    }
                    disabled={false}
                    inaccessibleTerrain={
                      Array [
                        "OCEAN",
                      ]
                    }
                    name="SOLDIER"
                    player="blue"
                    stats={
                      Object {
                        "attack": 0.4,
                        "defense": 4,
                        "move": 3,
                        "range": 1,
                      }
                    }
                    type="LAND"
                  />
                  <Health
                    className="unit"
                    health={20}
                  />
                </div>
              </div>
            </Grass>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-4-2-SAND"
        >
          <Connect(Terrain)
            column={4}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={2}
            terrain={
              Object {
                "defenseBonus": 1,
              }
            }
          >
            <Sand
              column={4}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={2}
              terrain={
                Object {
                  "defenseBonus": 1,
                }
              }
            >
              <div />
            </Sand>
          </Connect(Terrain)>
        </td>
      </tr>
      <tr
        className="matrix-row"
        key="matrix-row-3"
      >
        <td
          className="matrix-cell"
          key="matrix-cell-0-3-LAND_FACTORY"
        >
          <Connect(Terrain)
            column={0}
            currentPlayer="BLUE"
            isCurrentPlayer={true}
            isOccupied={false}
            isOccupiedByPlayer={false}
            player="BLUE"
            row={3}
            terrain={
              Object {
                "actionable": true,
                "defenseBonus": 1,
                "selectable": true,
              }
            }
          >
            <LandFactory
              column={0}
              currentPlayer="BLUE"
              isCurrentPlayer={true}
              isOccupied={false}
              isOccupiedByPlayer={false}
              player="BLUE"
              row={3}
              terrain={
                Object {
                  "actionable": true,
                  "defenseBonus": 1,
                  "selectable": true,
                }
              }
            >
              <div>
                <Health
                  className="building"
                  health={20}
                />
              </div>
            </LandFactory>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-1-3-SAND"
        >
          <Connect(Terrain)
            column={1}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={3}
            terrain={
              Object {
                "defenseBonus": 1,
              }
            }
          >
            <Sand
              column={1}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={3}
              terrain={
                Object {
                  "defenseBonus": 1,
                }
              }
            >
              <div />
            </Sand>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-2-3-SAND"
        >
          <Connect(Terrain)
            column={2}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={3}
            terrain={
              Object {
                "defenseBonus": 1,
              }
            }
          >
            <Sand
              column={2}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={3}
              terrain={
                Object {
                  "defenseBonus": 1,
                }
              }
            >
              <div />
            </Sand>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-3-3-GRASS"
        >
          <Connect(Terrain)
            column={3}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={true}
            isOccupiedByPlayer={true}
            row={3}
            terrain={
              Object {
                "defenseBonus": 0,
              }
            }
          >
            <Grass
              column={3}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={true}
              isOccupiedByPlayer={true}
              row={3}
              terrain={
                Object {
                  "defenseBonus": 0,
                }
              }
            >
              <div>
                <div>
                  <Soldier
                    attacks={
                      Array [
                        "AIR",
                        "LAND",
                        "SEA",
                      ]
                    }
                    disabled={false}
                    inaccessibleTerrain={
                      Array [
                        "OCEAN",
                      ]
                    }
                    name="SOLDIER"
                    player="blue"
                    stats={
                      Object {
                        "attack": 0.4,
                        "defense": 4,
                        "move": 3,
                        "range": 1,
                      }
                    }
                    type="LAND"
                  />
                  <Health
                    className="unit"
                    health={20}
                  />
                </div>
              </div>
            </Grass>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-4-3-SAND"
        >
          <Connect(Terrain)
            column={4}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={3}
            terrain={
              Object {
                "defenseBonus": 1,
              }
            }
          >
            <Sand
              column={4}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={3}
              terrain={
                Object {
                  "defenseBonus": 1,
                }
              }
            >
              <div />
            </Sand>
          </Connect(Terrain)>
        </td>
      </tr>
      <tr
        className="matrix-row"
        key="matrix-row-4"
      >
        <td
          className="matrix-cell"
          key="matrix-cell-0-4-SAND"
        >
          <Connect(Terrain)
            column={0}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={4}
            terrain={
              Object {
                "defenseBonus": 1,
              }
            }
          >
            <Sand
              column={0}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={4}
              terrain={
                Object {
                  "defenseBonus": 1,
                }
              }
            >
              <div />
            </Sand>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-1-4-GRASS"
        >
          <Connect(Terrain)
            column={1}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={4}
            terrain={
              Object {
                "defenseBonus": 0,
              }
            }
          >
            <Grass
              column={1}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={4}
              terrain={
                Object {
                  "defenseBonus": 0,
                }
              }
            >
              <div />
            </Grass>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-2-4-SAND"
        >
          <Connect(Terrain)
            column={2}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={true}
            isOccupiedByPlayer={true}
            row={4}
            terrain={
              Object {
                "defenseBonus": 1,
              }
            }
          >
            <Sand
              column={2}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={true}
              isOccupiedByPlayer={true}
              row={4}
              terrain={
                Object {
                  "defenseBonus": 1,
                }
              }
            >
              <div>
                <div>
                  <SmallTank
                    attacks={
                      Array [
                        "AIR",
                        "LAND",
                        "SEA",
                      ]
                    }
                    disabled={false}
                    inaccessibleTerrain={
                      Array [
                        "OCEAN",
                        "MOUNTAIN",
                      ]
                    }
                    name="SMALL_TANK"
                    player="blue"
                    stats={
                      Object {
                        "attack": 0.8,
                        "defense": 6,
                        "move": 4,
                        "range": 1,
                      }
                    }
                    type="LAND"
                  />
                  <Health
                    className="unit"
                    health={20}
                  />
                </div>
              </div>
            </Sand>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-3-4-SAND"
        >
          <Connect(Terrain)
            column={3}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={false}
            isOccupiedByPlayer={false}
            row={4}
            terrain={
              Object {
                "defenseBonus": 1,
              }
            }
          >
            <Sand
              column={3}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={false}
              isOccupiedByPlayer={false}
              row={4}
              terrain={
                Object {
                  "defenseBonus": 1,
                }
              }
            >
              <div />
            </Sand>
          </Connect(Terrain)>
        </td>
        <td
          className="matrix-cell"
          key="matrix-cell-4-4-SAND"
        >
          <Connect(Terrain)
            column={4}
            currentPlayer="BLUE"
            isCurrentPlayer={false}
            isOccupied={true}
            isOccupiedByPlayer={false}
            row={4}
            terrain={
              Object {
                "defenseBonus": 1,
              }
            }
          >
            <Sand
              column={4}
              currentPlayer="BLUE"
              isCurrentPlayer={false}
              isOccupied={true}
              isOccupiedByPlayer={false}
              row={4}
              terrain={
                Object {
                  "defenseBonus": 1,
                }
              }
            >
              <div>
                <div>
                  <Soldier
                    attacks={
                      Array [
                        "AIR",
                        "LAND",
                        "SEA",
                      ]
                    }
                    disabled={false}
                    inaccessibleTerrain={
                      Array [
                        "OCEAN",
                      ]
                    }
                    name="SOLDIER"
                    player="yellow"
                    stats={
                      Object {
                        "attack": 0.4,
                        "defense": 4,
                        "move": 3,
                        "range": 1,
                      }
                    }
                    type="LAND"
                  />
                  <Health
                    className="unit"
                    health={20}
                  />
                </div>
              </div>
            </Sand>
          </Connect(Terrain)>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
